src/optics.jl

```diff
@@ -229,50 +278,106 @@ julia> using Accessors
 
 julia> obj = (a=1, b=2);
 
-julia> Accessors.mapproperties(x -> x+1, obj)
+julia> Accessors.mapobject(x -> x+1, obj)
 (a = 2, b = 3)
 ```
 $EXPERIMENTAL
 """
-function mapproperties(f, obj)
+function mapobject(f, obj::O, ::Properties, handler) where O
     # TODO move this helper elsewhere?
-    # TODO should we use a generated function based on fieldnames?
     pnames = propertynames(obj)
     if isempty(pnames)
-        return obj
+        return skip(handler) ? () : obj
     else
-        ctor = constructorof(typeof(obj))
-        new_props = map(pnames) do p
+        ctr = _constructor(handler, O)
+        args = map(pnames) do p
             f(getproperty(obj, p))
         end
-        return ctor(new_props...)
+        return ctr(args...)
+    end
+end
+function mapobject(f, obj::O, ::Properties, handler, itr::Int) where O
+    pnames = propertynames(obj)
+    if isempty(pnames)
+        return obj, itr
+    else
+        # TODO: this is too slow
+        new_props, itr = reduce(pnames; init=((), itr)) do (vals, itr), p
+            val, itr = f(getproperty(obj, p), itr)
+            (vals..., val), itr
+        end
+        ctr = _constructor(handler, O)
+        return ctr(new_props...), itr
     end
 end
 
 """
-    Properties()
+    Fields()
 
-Access all properties of an objects.
+Access all fields of an objects.
 
 ```jldoctest
 julia> using Accessors
 
 julia> obj = (a=1, b=2, c=3)
 (a = 1, b = 2, c = 3)
 
-julia> set(obj, Properties(), "hi")
+julia> set(obj, Fields(), "hi")
 (a = "hi", b = "hi", c = "hi")
 
-julia> modify(x -> 2x, obj, Properties())
+julia> modify(x -> 2x, obj, Fields())
 (a = 2, b = 4, c = 6)
 ```
-Based on [`mapproperties`](@ref).
+Based on [`mapfields`](@ref).
 
 $EXPERIMENTAL
 """
-struct Properties end
```

Comment:
Since we have `ConstructionBase.getproperties` and `ConstructionBase.setproperties` I wonder if we could base this on
```julia
set(obj, ::Properties, val) = setproperties(obj, val)
(::Properties)(obj) = getproperties(obj)
```

Reply:
Yeah that's probably better.